### ISSUE TITLE
Staggered eigensolver gives wrong smallest eigenvalue when splitting a 16^4 lattice in `t` direction

### ISSUE THREAD

**Issue by weinbe2** [lattice/quda#877]
I don't think the issue is in the staggered eigensolver specifically, it's just the only place I've noticed an issue (so far). This issue is unrelated to the zero mass quirk I recently addressed, as it shows up before and after the fix. I've only verified it exists on `release/1.0.x` (and my hotfix branch). Reference commands given at the very bottom.

Overly verbose description of what's going on and what I've tried:

When running the eigensolver, with or without polynomial acceleration, I find a different answer if I run on two GPUs, splitting in the `t` direction, than if I run on a single GPU or on two GPUs splitting in the, say, `z` direction instead. The single GPU and two GPUs splitting in `z` results are "truly correct" as verified by runs on independent code by Craig McNeil on a given configuration.

I can also reproduce this error if I split the lattice further, i.e., running on eight GPUs, splitting in `y`, `z`, and `t`. The wrong answer I get in this case is consistent with just splitting in the `t` direction. This bug is independent of running in row or column major (which wouldn't matter in the two GPU case anyway, but could factor in for whatever reason in the eight GPU case).

I cannot reproduce this if I run on a single GPU and use `--partition 8` instead.

Similarly, I can't reproduce any issue while running `staggered_dslash_ctest`. 

Unfortunately, I can't seem to reproduce it on a random gauge field (I get different answers for all partitionings), so for reference here's a link to Craig's field (hopefully it stays there for a while). It's a 16^4 lattice so it's only ~36 MB: https://www.dropbox.com/s/cdhpze3thsiiwuw/Q4800U_ildg.000700?dl=0

I've verified that I get consistent (right or wrong) answers with or without polynomial acceleration, so it's not an issue related to eigenvalues "falling through the cracks."

Things I haven't tried yet (for time reasons):
* Seeing a difference in iteration count when running CG instead.
* Disabling GDR or P2P to see if it's policy dependent (GDR unlikely to have anything to do with it)

If anyone has the chance to take the torch on debugging this a bit, or at least seeing if they can reproduce it themselves, I'd be grateful. This was run on two V100's connected over P2P.

Reference command which gives the wrong answer:
```
mpirun -np 2 ./staggered_eigensolve_test --dim 16 16 16 8 --gridsize 1 1 1 2 --load-gauge ./Q4800U_ildg.000700 --dslash-type asqtad --mass 0.01 --compute-fat-long true --tadpole-coeff 1.0 --test 3 --prec double --recon 18 --eig-nConv 16 --eig-nEv 18 --eig-nKr 32 --eig-tol 1e-7 --eig-use-normop false --eig-use-poly-acc false --eig-spectrum SR "
```

To get the right answer, replace the first bit of that with either:
```
mpirun -np 2 ./staggered_eigensolve_test --dim 16 16 8 16 --gridsize 1 1 2 1 [...]
```
or
```
mpirun -np 1 ./staggered_eigensolve_test --dim 16 16 16 16 --gridsize 1 1 1 1 [...]
```
Technically I'm running this through Kate's binding scripts, but this is the reduced command.

**Comment by cpviolator** [comment]
Updating this issue with some tests from `staggered_invert_test`

Baseline command, one MPI rank:

```
mpirun -np 1 ./staggered_invert_test --dim 16 16 16 16 --gridsize 1 1 1 1 --load-gauge /scratch/CPviolator/work/Q4800U_ildg.000700 --dslash-type asqtad --mass 0.01 --compute-fat-long true --tadpole-coeff 1.0 --test 3 --prec double --recon 18 --niter 1000
```

gives

```
CG: 746 iterations, <r,r> = 5.801168e-10, |r|/|b| = 9.400380e-08
CG: Reliable updates = 7
CG: Convergence at 746 iterations, L2 relative residual: iterated = 9.400380e-08, true = 9.400380e-08 (requested = 1.000000e-07)
Solution = 4.79924e+06
Reconstructed: CUDA solution = 4.79924e+06, CPU copy = 4.79924e+06
0.988028 0.988028
Residuals: (L2 relative) tol 1e-07, QUDA = 9.40038e-08, host = 9.40038e-08; (heavy-quark) tol 0, QUDA = 2.01788e-08, host = 2.01788e-08
done: total time = 0.348028 secs, compute time = 0.356613 secs, 746 iter / 0.356613 secs = 164.422 gflops, 
```

But with a `t` splitting command:

```
mpirun -np 2 ./staggered_invert_test --dim 16 16 16 8 --gridsize 1 1 1 2 --load-gauge /scratch/CPviolator/work/Q4800U_ildg.000700 --dslash-type asqtad --mass 0.01 --compute-fat-long true --tadpole-coeff 1.0 --test 3 --prec double --recon 18 --niter 1000
```

we see

```
CG: 839 iterations, <r,r> = 6.092762e-10, |r|/|b| = 9.633738e-08
CG: Reliable updates = 8
CG: Convergence at 839 iterations, L2 relative residual: iterated = 9.633738e-08, true = 9.633738e-08 (requested = 1.000000e-07)
Solution = 3.31144e+06
Reconstructed: CUDA solution = 3.31144e+06, CPU copy = 3.31144e+06
0.988028 0.988028
Residuals: (L2 relative) tol 1e-07, QUDA = 9.63374e-08, host = 9.63374e-08; (heavy-quark) tol 0, QUDA = 1.9283e-08, host = 1.9283e-08
done: total time = 0.449209 secs, compute time = 0.463005 secs, 839 iter / 0.463005 secs = 142.449 gflops, 
```

Both done with `QUDA_ENABLE_TUNING=0`

**Comment by cpviolator** [comment]
Possibly due to the refactoring of the staggered gauge utils :( I'll retest on an old commit.

**Comment by cpviolator** [comment]
Did a few more tests.

* The eigensolver works fine for Wilson fermions.
* An old version of `staggered_invert_test` with the old gauge utils produces the same errors as above.
* Same errors appear when using `--dslash-type staggered` and/or `--compute-fat-long true/false`

**Comment by cpviolator** [comment]
Big Clue!

* The temporal periodicity is hardcoded in both `staggered_invert_test` and `staggered_eigensolve_test` to `QUDA_ANTI_PERIODIC_T`. Switching to `QUDA_PERIODIC_T` makes the issue go away in both the invert test and eigensolve test in that all the different MPI configurations give the same data.

Conclusion: temporal boundary condition not being properly implemented for staggered T comms.

**Comment by cpviolator** [comment]
Fixed with #878
